spec/lib/rmt/deduplicator_spec.rb

```diff
@@ -14,83 +20,149 @@
+
+          expect(DownloadedFile.where(local_path: dest_path).count).to eq(0)
+        end
       end
     end
 
     context 'copy' do
       before do
         deduplication_method(:copy)
-        add_downloaded_file(checksum_type, checksum, source_path)
-        deduplicate(checksum_type, checksum, dest_path)
+        add_downloaded_file(file.checksum_type, file.checksum, source_path)
       end
 
-      it('duplicates file') { expect(File.read(dest_path)).to eq(File.read(source_path)) }
-      it('duplicated file with copy') { expect(File.stat(source_path).nlink).to eq(1) }
-    end
+      context 'when there is a valid source file' do
+        it 'duplicates file without hardlink' do
+          deduplicate
 
-    context 'copy with changed file' do
-      before do
-        deduplication_method(:copy)
-        add_downloaded_file(checksum_type, 'foo', source_path)
-        File.open(source_path, 'a') { |f| f.puts 'this is a change' }
-        deduplicate(checksum_type, 'foo', dest_path)
+          expect(File.read(dest_path)).to eq(File.read(source_path))
+          expect(File.stat(source_path).nlink).to eq(1)
+        end
       end
 
-      it('duplicates file') { expect(File.exist?(dest_path)).to be_falsey }
-      it('source is not linked') { expect(File.stat(source_path).nlink).to eq(1) }
-    end
+      context 'when there is not a valid source file' do
+        before do
+          deduplication_method(:copy)
+          add_downloaded_file(file.checksum_type, 'foo', source_path)
+          File.open(source_path, 'a') { |f| f.puts 'this is a change' }
+        end
 
-    context 'hardlink with proper checksum' do
-      before do
-        deduplication_method(:hardlink)
-        add_downloaded_file(checksum_type, checksum, source_path)
-        deduplicate(checksum_type, checksum, dest_path)
-      end
+        let(:checksum) { 'foo' }
 
-      it('duplicates file') { expect(File.read(dest_path)).to eq(File.read(source_path)) }
-      it('duplicated file with copy') { expect(File.stat(source_path).nlink).to eq(2) }
-    end
+        it('returns false') { is_expected.to be false }
 
-    context 'hardlink not supported' do
-      before do
-        deduplication_method(:hardlink)
-        add_downloaded_file(checksum_type, checksum, source_path)
-        allow(::FileUtils).to receive(:ln).and_raise(StandardError)
-      end
+        it 'does not duplicate file' do
+          deduplicate
 
-      it('throws hardlink exception') do
-        expect { deduplicate(checksum_type, checksum, dest_path) }.to raise_error(::RMT::Deduplicator::HardlinkException)
+          expect(File.exist?(dest_path)).to be false
+        end
+
+        it 'removes invalid source files' do
+          deduplicate
+
+          expect(File.exist?(source_path)).to be false
+        end
+
+        it 'untracks invalid source files' do
+          expect { deduplicate }
+            .to change { DownloadedFile.where(local_path: source_path).count }
```

Comment:
@lcaparroz consistent, you say? :slightly_smiling_face:

Reply:
Specs don't have problems with module levels, as you would expect in a regular project file, you usually don't need the `::` to avoid namespace conflicts. I could say I've been consistent at two different levels: tests, which don't need `::` at all, and application files, [where it might be a good idea to be explicit about namespaces](https://github.com/rubocop-hq/ruby-style-guide#namespace-definition).

Reply:
Yeah, so you are not being consistent.
Being consistent would mean referencing every single class starting from the root namespace -- this is just not a common development practice.
Besides, why would an application need it, but not the tests? Neither one needs it.

Reply:
I removed the `::` operators from files I changed in this PR.